crypto/x509/by_store.c

```diff
@@ -116,11 +116,25 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
     case X509_L_ADD_STORE:
         if (argp != NULL) {
             STACK_OF(OPENSSL_STRING) *uris = X509_LOOKUP_get_method_data(ctx);
-            char *data = OPENSSL_strdup(argp);
+            char *data;
+            const char *scheme_end = argp;
 
-            if (data == NULL) {
-                return 0;
+            /*
+             * for file URIs and URIs without a scheme (which implies "file:"),
+             * immediately check if the referenced file exists
+             */
+            OSSL_SKIP_SCHEME(scheme_end);
+            if (HAS_CASE_PREFIX(argp, "file:")
+                || scheme_end == argp || *scheme_end != ':') {
+                struct stat st;
+
+                if (OSSL_file_stat(argp, &st) == NULL)
+                    return 0;
+                /* likely not worth checking in addition that file is readable */
             }
```

Comment:
Case study: what happens if I have a file `foo:bar.pem`?  Or you know, if some cert issuer gives you a file with a name derived from the issue and subject with a colon in between, like oh-I-dunno, `GoogleCa:RichardsServer.pem`?

I know we're playing on the edge of commonality here, but I for one have sometimes downloaded files that include a colon in their names, and some of them could legitimately be seen as URIs.

Reply:
This should work just fine.
I just gave it a try, and it does:

```
cp test/certs/root-cert.pem foo:cert.pem
apps/openssl verify -CAstore foo:cert.pem test/certs/root-cert.pem 
```
yields
```
test/certs/root-cert.pem: OK
```
which implies the filename got accepted and loaded successfully.

I'm going to add also this and some further cases to `25-test_verify.t`

Reply:
... or for that matter, on systems that have symbolic devices, `CAs:blargh.pem` is 100% legitimate, and isn't even weird.

Reply:
> This should work just fine. I just gave it a try, and it does:
> 
> ```
> cp test/certs/root-cert.pem foo:cert.pem
> apps/openssl verify -CAstore foo:cert.pem test/certs/root-cert.pem 
> ```
> 
> yields
> 
> ```
> test/certs/root-cert.pem: OK
> ```

If I don't copy (i.e. foo:cert.pem doesn't exist), I get:

``` console
$ ./util/wrap.pl apps/openssl verify -CAstore foo:cert.pem test/certs/root-cert.pem 
CN=Root CA
error 18 at 0 depth lookup: self-signed certificate
error test/certs/root-cert.pem: verification failed
```

Reply:
That's on purpose - see new the code in `by_store.c`.
At places where a URI may be used, anything like `foo:cert.pem` where `foo` is a legitimate scheme unequal to `file`,
we must expect that this is not a file and hope that, on demand, there will be any provider that can deal with it.

It we want cover this as well with an early error, we'll have to add a generic provider API function for checking the existence of a credential given its URI - which I guess is what you were alluding to in https://github.com/openssl/openssl/issues/27461#issuecomment-2826600425.